sources/csv/data.go

```diff
@@ -0,0 +1,284 @@
+		}
+		count++
+	}
+	// Exclude the first row, that is, column headers.
+	return count - 1, nil
+}
+
+// ProcessCSV writes data across the tables provided in the manifest file. Each table's data can be provided
+// across multiple CSV files hence, the manifest accepts a list of file paths in the input.
+func ProcessCSV(conv *internal.Conv, tables []Table) error {
+	for _, table := range tables {
+		// Populating just the table names in the conv for SrcSchema and SpSchema
+		// so the report for row stats is generated.
+		conv.SrcSchema[table.Table_name] = schema.Table{Name: table.Table_name}
+
+		// The map colDefs stores the mapping from column names to their final types.
+		colDefs := make(map[string]ddl.ColumnDef)
+		for _, col := range table.Columns {
+			ty, err := ToSpannerType(col.Type_name)
+			if err != nil {
+				return fmt.Errorf("can't map to spanner type: %v. Please use the data types as in your spanner database", err)
+			}
+			colDefs[col.Column_name] = ddl.ColumnDef{Name: col.Column_name, T: ty}
+		}
+		conv.SpSchema[table.Table_name] = ddl.CreateTable{Name: table.Table_name, ColDefs: colDefs}
+
+		for _, filePath := range table.File_patterns {
+			csvFile, err := os.Open(filePath)
+			if err != nil {
+				return fmt.Errorf(fmt.Sprintf("can't read csv file: %s due to: %v\n", filePath, err))
+			}
+			r := csvReader.NewReader(csvFile)
+
+			// First row is expected to be the column headers.
+			srcCols, err := r.Read()
+			if err == io.EOF {
+				conv.Unexpected(fmt.Sprintf("File %s is empty.", filePath))
+				continue
+			}
+			if err != nil {
+				return fmt.Errorf("can't read csv headers for col names due to: %v", err)
+			}
+			for {
+				values, err := r.Read()
+				if err == io.EOF {
+					break
+				}
+				if err != nil {
+					return fmt.Errorf(fmt.Sprintf("can't read row  names due to: %v", err))
+				}
+				processDataRow(conv, table.Table_name, srcCols, values)
+			}
+		}
+	}
+	return nil
+}
+
+// processDataRow converts a row into go data types as per the client libs.
+func processDataRow(conv *internal.Conv, tableName string, srcCols []string, values []string) {
+	cvtVals, err := convertData(conv, tableName, srcCols, values)
+	if err != nil {
+		conv.Unexpected(fmt.Sprintf("Error while converting data: %s\n", err))
+		conv.StatsAddBadRow(tableName, conv.DataMode())
+		conv.CollectBadRow(tableName, srcCols, values)
+	} else {
+		conv.WriteRow(tableName, tableName, srcCols, cvtVals)
+	}
+}
+
+// convertData currently only supports scalar data types.
+func convertData(conv *internal.Conv, tableName string, srcCols []string, values []string) ([]interface{}, error) {
+	var v []interface{}
+	colDefs := conv.SpSchema[tableName].ColDefs
+	for i, val := range values {
+		colName := srcCols[i]
+		x, err := convScalar(colDefs[colName].T, val)
+		if err != nil {
+			return nil, err
+		}
+		v = append(v, x)
+	}
+	return v, nil
+}
+
+func convScalar(spannerType ddl.Type, val string) (interface{}, error) {
```

Comment:
Is it possible to have array as datatype too ?

Reply:
Yes I was planning to take that during the next cut.